### PR TITLE
Minimize to tray with libindicator

### PR DESCRIPTION
**Pull request by len-foss**
Hello,

First, thank you very much for this wonderful program. 
It is very nice to be able to use a secure device for security purposes.

Some sites are very annoying with MFA, and require multiple entries. 
I tend to prefer having it minimized to tray for these cases. 
In terms of security, I think it does not really pose a problem since there are good auto lock on system lock and inactive settings that are available already.

Although I think if it were to be accepted, it would make sense to add a setting for it and disable it by default.

Let me know what you think!

### CODE REVIEW

**Comment by paolostivanin** [comment]
Hello,
thanks a lot for your contribution, very much appreciated :smile: 
Let's discuss some points:
1. yes, we would need to add an option to **enable** such behavior (should be disabled by default)
2. would it work on other DE too? (e.g. KDE)
3. would it work on Wayland? If not, we would need to deal with that at runtime

**Comment by len-foss** [comment]
I agree with point 1. For point 2, it does work on KDE, and since that's what I'm using it will be the most thoroughly tested.

Wayland seems to open a can of worms though.
It seems that the best way is to implement it via StatusNotifierItem (SNI), which if I understand would work if the user installs https://github.com/ubuntu/gnome-shell-extension-appindicator. 

Should I try to implement it using this standard instead?

**Comment by paolostivanin** [comment]
I'm sorry, I don't have enough knowledge on tray icon standards & co to give a real answer.
It does seem that using, for example, https://github.com/AyatanaIndicators/libayatana-appindicator could work.

**Comment by len-foss** [comment]
You're right, thank you! It seems that I read outdated doc on the subject.

I've tested the following on KDE X11 and Wayland, and everything works as expected.

Everything should be ready for review I think.

**Comment by paolostivanin** [comment]
Thank you very much! I'll try to give you a feedback by end of week :smile:

File: CMakeLists.txt [hunk 1]
```diff
@@ -47,7 +47,7 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
 endif()
 
 add_definitions(-DGETTEXT_PACKAGE=\"${GETTEXT_PACKAGE}\")
-
+add_definitions(-DDESTINATION="${CMAKE_INSTALL_PREFIX}")
```
Comment (paolostivanin) [inline CMakeLists.txt:50]: we can merge those two with:

```
add_definitions(
    -DGETTEXT_PACKAGE=\"${GETTEXT_PACKAGE}\"
    -DDESTINATION=\"${CMAKE_INSTALL_PREFIX}\"
)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -70,6 +70,7 @@ pkg_check_modules(UUID REQUIRED uuid>=2.34.0)
 pkg_check_modules(PROTOC REQUIRED libprotobuf-c>=1.3.0)
 pkg_check_modules(LIBSECRET REQUIRED libsecret-1>=0.20.0)
 pkg_check_modules(LIBQRENCODE REQUIRED libqrencode>=4.0.2)
+pkg_check_modules(APPINDICATOR REQUIRED ayatana-appindicator3-0.1)
```
Comment (paolostivanin) [inline CMakeLists.txt:73]: I'm still unsure about whether this should be mandatory or not. I think distro packagers should be able to ship otpclient without the tray icon feature (for whatever reason).

**Comment by paolostivanin** [comment]
additional todo: please rebase your commits and sign the final one :smile:  thanks a lot!

**Comment by len-foss** [comment]
I have added an option to skip the dependency altogether, and a test to run the install without it.
To avoid cluttering the code with too much `#ifdef`, the settings do exist even when the option is not there, they are just unplugged. 
Because in a GtkGrid it isn't possible to hide child elements individually, it seemed much simpler to remove the row.
The other cmake lines referring to APPINDICATOR_LIBRARIES do not seem to matter either, so they are not between `if` either.

I think this approach keeps the complexity minimal enough.

**Comment by len-foss** [comment]
By the way, I had a suggestion to update https://github.com/paolostivanin/OTPClient/wiki/Secure-Memory-Limitations#how-to-set-memlock

```
* then, locate if a file containing the limits exists with `grep -rni memlock /etc/security/`.
If none exists, you can create a file called, for example, `/etc/security/limits.d/memlock.conf`. 
Then add the following text:
```
It can be the case that there is already a file even if it was not set up.

**Comment by paolostivanin** [comment]
thanks a lot for your work! I'll test it in the coming days on GNOME and KDE too and then, if all good, I will merge it.

PS wiki updated, thank you

File: src/gui/settings-cb.c [hunk 3]
```diff
@@ -86,6 +95,14 @@ settings_dialog_cb (GSimpleAction *simple UNUSED,
     GtkWidget *dt_switch = GTK_WIDGET(gtk_builder_get_object (builder, "dark_theme_switch_id"));
     settings_data->dss_switch = GTK_WIDGET(gtk_builder_get_object (builder, "secret_service_switch_id"));
     g_signal_connect (settings_data->dss_switch, "state-set", G_CALLBACK(handle_autolock), settings_data);
+    settings_data->tray_switch = GTK_WIDGET(gtk_builder_get_object (builder, "tray_switch_id"));
+    #ifdef ENABLE_MINIMIZE_TO_TRAY
+    g_signal_connect (settings_data->tray_switch, "state-set", G_CALLBACK(handle_tray_switch), settings_data);
+    #endif
```
Comment (paolostivanin) [inline src/gui/settings-cb.c:101]: we could use and `#else` here

**Comment by len-foss** [comment]
Thank you for talking the time, and don't hesitate it you want more changes to be done. I think I have approached most of it to be as simple as possible and limit the amount of potential bugs (for instance, destroying and recreating the indicator at each setting change seemed janky, whereas hiding it will basically have the same effect at next restart).

**Comment by paolostivanin** [comment]
indeed! I really do like your implementation :smile: 
Hopefully, I will be able to test everything and merge it before end of this week!

**Comment by paolostivanin** [comment]
@len-foss could you please rebase into a single commit? Thanks :smile:

**Comment by len-foss** [comment]
Done!

**Comment by joakim-tjernlund** [comment]
On MATE DE and also like this addition.

**Comment by paolostivanin** [comment]
Thanks a lot for your contribution! Let's merge this into master :smile: I'll make a new release in a few weeks. I need to tackle some flatpak bugs first.

**Comment by len-foss** [comment]
Nice! Don't hesitate to ping me if there are any bugs reported when there are more users with different setups.

**Comment by paolostivanin** [comment]
fantastic! Thanks a lot again :smile:

**Pull request merged**